customerio/client_base.py

```diff
@@ -0,0 +1,77 @@
+"""
+Implements the base client that is used by other classes to make requests
+"""
+from __future__ import division
+from datetime import datetime
+import math
+import time
+
+from requests import Session
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.util.retry import Retry
+
+try:
+    from datetime import timezone
+    USE_PY3_TIMESTAMPS = True
+except ImportError:
+    USE_PY3_TIMESTAMPS = False
+
+class CustomerIOException(Exception):
+    pass
+
+class ClientBase(object):
+    def __init__(self, retries=3, timeout=10, backoff_factor=0.02):
+        self.timeout = timeout
+        self.retries = retries
+
+        self.http = Session()
+        # Retry request a number of times before raising an exception
+        # also define backoff_factor to delay each retry
+        self.http.mount('https://', HTTPAdapter(max_retries=Retry(
+            total=retries, backoff_factor=backoff_factor)))
+
+    def send_request(self, method, url, data):
+        '''Dispatches the request and returns a response'''
+
+        try:
+            response = self.http.request(
+                method, url=url, json=self._sanitize(data), timeout=self.timeout)
+        except Exception as e:
+            # Raise exception alerting user that the system might be
+            # experiencing an outage and refer them to system status page.
+            message = '''Failed to receive valid reponse after {count} retries.
+Check system status at http://status.customer.io.
+Last caught exception -- {klass}: {message}
+            '''.format(klass=type(e), message=e, count=self.retries)
+            raise CustomerIOException(message)
+
+        result_status = response.status_code
+        if result_status != 200:
+            raise CustomerIOException('%s: %s %s' % (result_status, url, data))
```

Comment:
We also should raise `response.text` otherwise we wouldn't show the user helpful error messages that the backend sends us. Something like:

```raise CustomerIOException('%s: %s %s. %s' % (result_status, url, data, response.text))```

@jatinn would this be okay, and not break any previous behavior?

Reply:
believe that should be fine as the exception has not changed.
one thing though is there always a helpful error message? and is the response always a string or is the error a nested key in json response?